wasm/src/decoder/util.rs

```diff
@@ -0,0 +1,50 @@
+use super::common;
+use crate::error::Error;
+use crate::error::Result;
+use leb128::read;
+
+pub(crate) fn decode_vec<T, R, F>(reader: &mut R, func: F) -> Result<Vec<T>>
+where
+    R: std::io::Read,
+    F: Fn(&mut R) -> Result<T>,
+{
+    let length = read::unsigned(reader)? as usize;
+    if length > u32::MAX as usize {
+        return Err(Error::ArrayTooLarge);
+    }
+
+    let mut v = Vec::with_capacity(length);
+
+    for x in v.iter_mut().take(length) {
+        *x = func(reader)?;
+    }
```

Comment:
Wouldn't this iterator be empty?

Reply:
clippy didn't like my previous solution and recommended this to me instead.

previous solution was below
```rs
let mut v = Vec::with_capacity(length);

for i in 0..length {
    v[i] = func(reader)?;
}
```

Reply:
It's because clippy lacks the knowledge that the Vec's *length is zero*.

This is just a bug, you want to push upon each iteration, not write into it via `*x =` or even `v[i] = `.

Reply:
```rust
for _ in 0..length {
 v.push(func(reader)?);
}
```

Reply:
oh yeah I see. Stupid mistake from myself. Weirdly enough I did do it correctly in the code section

```rs
for _ in 0..x.0 {
            locals.push(x.1);
        }
        ```